torch/tensortypes/nestedtensor/nested.py

```diff
@@ -0,0 +1,215 @@
+                raise ValueError("Each element of the tuple or list must "
+                                 "be a torch.Tensor")
+        tensors = []
+        for data_ in data:
+            # torch.tensor copies on construction
+            new_data = torch.empty_like(data_)
+            new_data.copy_(data_)
+            new_data = new_data.to(dtype)
+            new_data = new_data.to(device)
+            new_data = new_data.requires_grad_(requires_grad)
+            if pin_memory:
+                new_data = new_data.pin_memory()
+            tensors.append(new_data)
+
+        return NestedTensor(tensors)
+
+def as_nestedtensor(data, dtype=None, device=None):
+    ret = NestedTensor(data)
+    if dtype is not None:
+        ret = ret.to(dtype)
+    if device is not None:
+        ret = ret.to(device)
+    return ret
+
+class NestedTensor():
+    # The attributes must match across all constiuents
+    # and default to the empty Tensor's if the given list
+    # is empty.
+    #
+    # The NestedTensor's attributes then become that of its
+    # constiuents.
+    #
+    # Attributes:
+    #     dim
+    #     layout
+    #     device
+    #     dtype
+    #     requires_grad
+    #     is_pinned
+    def __init__(self, tensors):
+        for tensor in tensors:
+            assert torch.is_tensor(tensor)
+        self.tensors = tensors
+        if len(tensors):
+            self.dim = tensors[0].dim()
+            self.layout = tensors[0].layout
+            self.device = tensors[0].device
+            self.dtype = tensors[0].dtype
+            requires_grad = tensors[0].requires_grad
+            is_pinned = tensors[0].is_pinned()
+            for tensor in tensors:
+                if not (self.dim == tensor.dim() and
+                        self.layout == tensor.layout and
+                        self.device == tensor.device and
+                        self.dtype == tensor.dtype and
+                        requires_grad == tensor.requires_grad and
+                        is_pinned == tensor.is_pinned()):
+                    raise ValueError("Each passed Tensor "
+                                     "must match in dim, layout, "
+                                     "device, dtype and requires_grad")
+        else:
+            # Carrying around information as member variables vs.
+            # checking one entry of the owned Tensors is annoying
+            # and error-prone. Carrying around an is_empty attribute
+            # to hide the fact that we carry around a list with a
+            # single empty Tensor is also annoying and error-prone.
+            # Both are not worth it for a minor feature.
+            raise ValueError("We do not support empty lists for now.")
+
+    def __getattribute__(self, attr):
+        if attr == 'shape':
+            raise NotImplementedError()
+        if attr == 'requires_grad':
+            return self.tensors[0].requires_grad
+        if attr == 'grad_fn':
```

Comment:
I don't think we let users use this in PyTorch?

Reply:
I couldn't find anywhere where we documented that `grad_fn` is callable. The implementation of `grad_fn` here is also kind of weird, because you'll have `nested_tensor.grad_fn != nested_tensor.grad_fn`

Reply:
It's part of torch.Tensor

Reply:
it's part of torch.Tensor, but iirc we don't allow users to call it or manipulate it.

Reply:
Well then why can they. Let's fix that.

Reply:
The logic for `grad_fn` is very old, and not documented because want to reserve the right to change the implementation details of it. It's also not something that most users every really need to work with.

Reply:
Ok, I disabled it.